### PR TITLE
Add NSX-T Firewall resource and data source

### LINKED ISSUE #622: Create new resource for NSX-T firewall management
NSX-T edge gateway firewall requires new resource and this task is to track this work.

### PR DESCRIPTION
**Pull request by Didainius**
Add `vcd_nsxt_firewall` resource and data source for managing NSX-T 

Fixes #622

### CODE REVIEW

File: .changes/v3.3.0/674-features.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+* **New Resource:** `vcd_nsxt_firewall` for NSX-T Edge Gateways [GH-663]
+* **New Data source:** `vcd_nsxt_firewall` for NSX-T Edge Gateways [GH-663]
```
Comment (lvirbalas) [inline .changes/v3.3.0/674-features.md:2]:
```suggestion
* **New Data Source:** `vcd_nsxt_firewall` for NSX-T Edge Gateways [GH-663]
```

File: vcd/datasource_vcd_nsxt_firewall.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			"edge_gateway_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Edge Gateway ID in which Firewall rules are located",
```
Comment (lvirbalas) [inline vcd/datasource_vcd_nsxt_firewall.go:32]:
```suggestion
				Description: "Edge Gateway ID in which Firewall Rules are located",
```

File: vcd/resource_vcd_nsxt_firewall.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			"edge_gateway_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Edge Gateway ID in which Firewall rule is located",
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall.go:47]: "Firewall rule" is a bit confusing here (and in the previous comment). As if we're talking about a single rule, but in fact we're talking about the whole firewall object. Maybe change "rule" -> "rules"? Or some other suggestions?

File: vcd/resource_vcd_nsxt_firewall.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+			"action":       value.Action,
+			"enabled":      value.Enabled,
+			"ip_protocol":  value.IpProtocol,
+			"direction":    value.Direction,
+			"logging":      value.Logging,
+			"sources":      sourceSet,
+			"destinations": destinationSet,
+			"applications": appPortProfileSet,
+		}
+
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall.go:284]: Extra line.

File: vcd/resource_vcd_nsxt_firewall.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			if oneRuleMapInterface["destinations"] != nil {
+				sourceGroups := convertSchemaSetToSliceOfStrings(oneRuleMapInterface["destinations"].(*schema.Set))
+				result[index].DestinationFirewallGroups = convertSliceOfStringsToOpenApiReferenceIds(sourceGroups)
+			}
+
+			if oneRuleMapInterface["applications"] != nil {
+				sourceGroups := convertSchemaSetToSliceOfStrings(oneRuleMapInterface["applications"].(*schema.Set))
+				result[index].ApplicationPortProfiles = convertSliceOfStringsToOpenApiReferenceIds(sourceGroups)
+			}
+
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall.go:322]: Hm, looks like one more extra line. gofmt missing?

File: vcd/resource_vcd_nsxt_firewall_test.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.#", "1"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.name", "test_rule"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.direction", "IN"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.ip_protocol", "IPV4"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.logging", "false"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.action", "ALLOW"),
+					resource.TestCheckResourceAttr("vcd_nsxt_firewall.testing", "rule.0.enabled", "true"),
+				),
+			},
+
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall_test.go:66]: Extra line? :)

File: vcd/resource_vcd_nsxt_firewall_test.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+const testAccVcdNsxtFirewallAndDs = testAccVcdNsxtFirewall + testAccVcdNsxtFirewallDS
+
+const testAccVcdNsxtFirewall2 = testAccVcdNsxtFirewallPrereqs + `
+resource "vcd_nsxt_security_group" "group" {
+  count = 3
+  org   = "{{.Org}}"
+  vdc   = "{{.NsxtVdc}}"
+
+  # Referring to a datasource for testing NSX-T Edge Gateway
+
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall_test.go:175]: Extra line :))

File: vcd/resource_vcd_nsxt_firewall_test.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+const testAccVcdNsxtFirewallAndDs = testAccVcdNsxtFirewall + testAccVcdNsxtFirewallDS
+
+const testAccVcdNsxtFirewall2 = testAccVcdNsxtFirewallPrereqs + `
+resource "vcd_nsxt_security_group" "group" {
+  count = 3
+  org   = "{{.Org}}"
+  vdc   = "{{.NsxtVdc}}"
+
+  # Referring to a datasource for testing NSX-T Edge Gateway
```
Comment (lvirbalas) [inline vcd/resource_vcd_nsxt_firewall_test.go:174]:
```suggestion
  # Referring to a data source for testing NSX-T Edge Gateway
```

File: website/docs/d/nsxt_firewall.html.markdown [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  edge gateway, you create firewall rules.
+---
+
+# vcd\_nsxt\_firewall
+
+Supported in provider *v3.3+* and VCD 10.1+ with NSX-T backed Edge Gateways.
+
+Provides a data source to read NSX-T Firewall configuration of an edge gateway. Firewalls allow 
+user to control the incoming and outgoing network traffic to and from an NSX-T Data Center 
+edge gateway, you create firewall rules.
```
Comment (lvirbalas) [inline website/docs/d/nsxt_firewall.html.markdown:17]: "you create firewall rules" ending seems abrupt. Please see if the paragraph can be clarified, rephrased or extended.

Nit: if we're capitalizing "Data Center", we should do same with "Edge Gateway".

File: website/docs/r/nsxt_firewall.html.markdown [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  Provides a resource to manage NSX-T Firewall. Firewalls allow user to control the incoming and 
+  outgoing network traffic to and from an NSX-T Data Center edge gateway, you create firewall rules.
+---
+
+# vcd\_nsxt\_firewall
+
+Supported in provider *v3.3+* and VCD 10.1+ with NSX-T backed Edge Gateways.
+
+Provides a resource to manage NSX-T Firewall. Firewalls allow user to control the incoming and 
+outgoing network traffic to and from an NSX-T Data Center edge gateway, you create firewall rules.
```
Comment (lvirbalas) [inline website/docs/r/nsxt_firewall.html.markdown:15]: Same ask as in the previous comment.

File: website/docs/r/nsxt_firewall.html.markdown [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+* `direction` - (Required) One of `IN`, `OUT`, or `IN_OUT`
+* `ip_protocol` - (Required) One of `IPV4`,  `IPV6`, or `IPV4_IPV6`
+* `enabled` - (Optional) Defines if the rule is enabled (default `true`)
+* `logging` - (Optional) Defines if logging for this rule is enabled (default `false`)
+* `action` - (Optional) Defines if it should `ALLOW` or `DROP` traffic (default `ALLOW`)
+* `sources` - (Optional) A set of source object Firewall Groups (`IP Sets` or `Security groups`). 
+Leaving it empty matches `Any` (all)
+* `destinations` - (Optional) A set of source object Firewall Groups (`IP Sets` or `Security groups`). 
+Leaving it empty matches `Any` (all)
+* `applications` - (Optional) A set of application port profiles. Leaving it empty matches  (all)
```
Comment (lvirbalas) [inline website/docs/r/nsxt_firewall.html.markdown:98]: The ending " (all)" seems abrupt.

**Review by lvirbalas — CHANGES_REQUESTED** [review]
A few asks to improve documentation, and some trivial nits, of course :)

File: .changes/v3.3.0/674-features.md (hunk 1 above)
Comment (Didainius) [reply .changes/v3.3.0/674-features.md:2]: Fixed.

File: vcd/datasource_vcd_nsxt_firewall.go (hunk 2 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_firewall.go]: Fixed

File: vcd/resource_vcd_nsxt_firewall.go (hunk 3 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall.go]: Yes, indeed. This was probably typo/leftover from initial version. Data Source had it right (plural). Fixed.

File: vcd/resource_vcd_nsxt_firewall.go (hunk 4 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall.go]: Removed

File: vcd/resource_vcd_nsxt_firewall.go (hunk 5 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall.go]: Nope. `gofmt` wouldn't complain about single line space. We have it in github tests. This is something that I leave as a separator while developing, but in fact leave it out when it is not needed anymore.. My bad.

File: vcd/resource_vcd_nsxt_firewall_test.go (hunk 6 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall_test.go]: Removed.

File: vcd/resource_vcd_nsxt_firewall_test.go (hunk 7 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall_test.go]: Ok, too many gaps here :) Removed.

File: vcd/resource_vcd_nsxt_firewall_test.go (hunk 8 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_firewall_test.go]: Fixed.

File: website/docs/d/nsxt_firewall.html.markdown (hunk 9 above)
Comment (Didainius) [reply website/docs/d/nsxt_firewall.html.markdown]: Yes. I actually copied some of these parts from docs and probably got confused. True about capitalization - I did that now.

File: website/docs/r/nsxt_firewall.html.markdown (hunk 10 above)
Comment (Didainius) [reply website/docs/r/nsxt_firewall.html.markdown]: Fixed. Removed that odd ending.

File: website/docs/r/nsxt_firewall.html.markdown (hunk 11 above)
Comment (Didainius) [reply website/docs/r/nsxt_firewall.html.markdown]: True. I have rephrased it to the same as in `sources` and `destinations` - hopefully this is better.

However - while re-reading this I saw these names again `source`, `destinations` and `applications`. What if I rename them to `source_ids`, `destination_ids` and `app_port_profile_ids`? In fact these are sets of IDs so probably this name makes more sense.

CC @dataclouder

Comment (lvirbalas) [reply website/docs/r/nsxt_firewall.html.markdown]: I like this suggestion! And we're using that approach in some places, which makes it explicitly clear:

Screenshot 2021-06-10 at 09 28 47

File: website/docs/r/nsxt_firewall.html.markdown [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+<a id="firewall-rule"></a>
+## Firwall Rule
+
+Each Firewall Rule contains following attributes:
+
+* `name` - (Required) Explanatory name for firewall rule (uniqueness not enforced)
+* `direction` - (Required) One of `IN`, `OUT`, or `IN_OUT`
+* `ip_protocol` - (Required) One of `IPV4`,  `IPV6`, or `IPV4_IPV6`
+* `enabled` - (Optional) Defines if the rule is enabled (default `true`)
```
Comment (lvirbalas) [inline website/docs/r/nsxt_firewall.html.markdown:92]: Would it be possible to have a test which tries to disable and then enable a rule? Though here it's unlikely, this scenario reminds me one issue in NSX-V which we had long time ago, so a test would be handy.

File: website/docs/r/nsxt_firewall.html.markdown [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+* `logging` - (Optional) Defines if logging for this rule is enabled (default `false`)
+* `action` - (Optional) Defines if it should `ALLOW` or `DROP` traffic (default `ALLOW`)
```
Comment (lvirbalas) [inline website/docs/r/nsxt_firewall.html.markdown:94]: I was looking at the examples and am hesitant about having a default (ALLOW) here. It's critical field and not seeing the actual value explicitly in HCL, makes it harder/riskier to use, IMHO. What do you and @dataclouder think? How does UI behave?

File: website/docs/r/nsxt_firewall.html.markdown (hunk 12 above)
Comment (Didainius) [reply website/docs/r/nsxt_firewall.html.markdown:92]: Added and tested.

File: website/docs/r/nsxt_firewall.html.markdown (hunk 13 above)
Comment (Didainius) [reply website/docs/r/nsxt_firewall.html.markdown:94]: UI sets `Allow` by default as there is this default rule "deny all"  in the end:
image

I am totally open to requiring it explicitly :) Not an issue maybe it enforces readability when checking without context.

Comment (lvirbalas) [reply website/docs/r/nsxt_firewall.html.markdown:94]: Unless there's some corner cases, which I'm not aware of, I vote for explicitly. We are doing that in other places where a binary choice exists in a critical field. To some extent, UI is also explicit in the sense that it is showing the Allow in the list.

Comment (Didainius) [reply website/docs/r/nsxt_firewall.html.markdown]: Implemented this in d17a124

**Review by lvirbalas — APPROVED** [review]
LGTM!

**Review by dataclouder — APPROVED** [review]

**Pull request merged**